File: Sources/SwiftUSD/Sdf/Sdf.swift
// Sdf.swift - Scene Description Foundation - Idiomatic Swift API
//
// This module provides Swift wrappers around CxxFacade's Sdf types,
// including paths, layers, asset paths, and layer offsets.
//
// The Sdf module is the foundation for scene description in USD,
// providing the core types for identifying objects (paths) and
// managing layer composition.

import CxxFacade

// MARK: - Type Aliases

/// SdfPath - Scene description path for identifying prims and properties.
///
/// SdfPath is used to specify a prim, property, or other element in a
/// scene graph. Paths are hierarchical and use forward slashes as separators.
///
/// ## Usage
/// ```swift
/// // Create paths
/// let rootPath = SdfPath("/")
/// let worldPath = SdfPath("/World")
/// let meshPath = SdfPath("/World/Mesh")
///
/// // Build paths hierarchically
/// let childPath = worldPath.AppendChild(TfToken("Child"))
///
/// // Property paths
/// let propertyPath = meshPath.AppendProperty(TfToken("visibility"))
/// ```
public typealias SdfPath = cxxfacade.sdf.SdfPath

/// SdfAssetPath - Asset path for referencing external files.
///
/// SdfAssetPath represents a path to an external asset (like a texture or
/// reference file). It can contain both the authored path and the resolved
/// path after asset resolution.
///
/// ## Usage
/// ```swift
/// // Create an asset path
/// let texturePath = SdfAssetPath("textures/diffuse.png")
///
/// // With resolved path
/// let resolvedPath = SdfAssetPath(
///     "textures/diffuse.png",
///     "/absolute/path/to/textures/diffuse.png"
/// )
/// ```
public typealias SdfAssetPath = cxxfacade.sdf.SdfAssetPath

/// SdfLayer - Scene description layer (USD file).
///
/// SdfLayer represents a single layer in the USD composition stack.
/// Each layer is typically backed by a file (usda, usdc, usd) and
/// contains scene description that can be composed with other layers.
///
/// ## Usage
/// ```swift
/// // Create an anonymous layer
/// let layer = SdfLayer.CreateAnonymous()
///
/// // Open an existing layer
/// let existingLayer = SdfLayer.FindOrOpen("scene.usda")
///
/// // Create a new layer at a path
/// let newLayer = SdfLayer.CreateNew("output.usda")
/// ```
public typealias SdfLayer = cxxfacade.sdf.SdfLayer

/// SdfLayerOffset - Time offset and scale between layers.
///
/// SdfLayerOffset is an affine transform (scale and translate) for time values.
/// It's used to remap animation timing when composing layers.
///
/// ## Usage
/// ```swift
/// // Create an identity offset (no change)
/// let identity = SdfLayerOffset()
///
/// // Create an offset that doubles time and shifts by 24 frames
/// var offset = SdfLayerOffset(24.0, 2.0)
///
/// // Apply offset to a time value
/// let remappedTime = offset * 10.0  // = 24 + 2*10 = 44
///
/// // Get the inverse transformation
/// let inverse = offset.GetInverse()
/// ```
public typealias SdfLayerOffset = cxxfacade.sdf.SdfLayerOffset

// MARK: - SdfPath Extensions

extension SdfPath: CustomStringConvertible {
    /// A textual representation of the path.
    public var description: String {
        return String(self.GetString())
    }
}

extension SdfPath: CustomDebugStringConvertible {
    /// A debug textual representation of the path.
    public var debugDescription: String {
        if self.IsEmpty() {
            return "SdfPath(empty)"
        }
        return "SdfPath(\"\(String(self.GetString()))\")"
    }
}

extension SdfPath: Hashable {
    /// Hashes the essential components of this value.
    public func hash(into hasher: inout Hasher) {
        hasher.combine(String(self.GetString()))
    }
}

extension SdfPath: ExpressibleByStringLiteral {
    /// Creates a path from a string literal.
    ///
    /// This allows you to write:
    /// ```swift
    /// let path: SdfPath = "/World/Mesh"
    /// ```
    public init(stringLiteral value: String) {
        self.init(value)
    }
}

extension SdfPath {
    /// Creates a path from a Swift String.
    public init(_ string: String) {
        self.init(std.string(string))
    }

    /// The path as a Swift String.
    public var pathString: String {
        return String(self.GetString())
    }

    /// Whether this path is empty.
    public var isEmpty: Bool {
        return self.IsEmpty()
    }

    /// Whether this is an absolute path (starts with /).
    public var isAbsolute: Bool {
        return self.IsAbsolutePath()
    }

    /// Whether this is a prim path (not a property path).
    public var isPrimPath: Bool {
        return self.IsPrimPath()
    }

    /// Whether this is a property path.
    public var isPropertyPath: Bool {
        return self.IsPropertyPath()
    }

    /// The parent path, or empty path if this is the root.
    public var parent: SdfPath {
        return self.GetParentPath()
    }

    /// The name (last element) of the path.
    public var name: TfToken {
        return self.GetName()
    }

    /// Append a child element to this path.
    public func appending(child: TfToken) -> SdfPath {
        return self.AppendChild(child)
    }

    /// Append a child element to this path using a string.
    public func appending(child: String) -> SdfPath {
        return self.AppendChild(TfToken(child))
    }

    /// Append a property to this path.
    public func appending(property: TfToken) -> SdfPath {
        return self.AppendProperty(property)
    }

    /// Append a property to this path using a string.
    public func appending(property: String) -> SdfPath {
        return self.AppendProperty(TfToken(property))
    }

    /// The absolute root path "/".
    public static var absoluteRoot: SdfPath {
        return SdfPath.AbsoluteRootPath()
    }

    /// An empty path.
    public static var empty: SdfPath {
        return SdfPath.EmptyPath()
    }
}

// MARK: - SdfAssetPath Extensions

extension SdfAssetPath: CustomStringConvertible {
    /// A textual representation of the asset path.
    public var description: String {
        return String(self.GetAssetPath())
    }
}

extension SdfAssetPath: CustomDebugStringConvertible {
    /// A debug textual representation of the asset path.
    public var debugDescription: String {
        let authored = String(self.GetAssetPath())
        let resolved = String(self.GetResolvedPath())
        if resolved.isEmpty {
            return "SdfAssetPath(\"\(authored)\")"
        }
        return "SdfAssetPath(\"\(authored)\" -> \"\(resolved)\")"
    }
}

extension SdfAssetPath: Hashable {
    /// Hashes the essential components of this value.
    public func hash(into hasher: inout Hasher) {
        hasher.combine(String(self.GetAssetPath()))
        hasher.combine(String(self.GetResolvedPath()))
    }
}

extension SdfAssetPath: ExpressibleByStringLiteral {
    /// Creates an asset path from a string literal.
    public init(stringLiteral value: String) {
        self.init(value)
    }
}

extension SdfAssetPath {
    /// Creates an asset path from a Swift String.
    public init(_ path: String) {
        self.init(std.string(path))
    }

    /// Creates an asset path with both authored and resolved paths.
    public init(_ path: String, resolved: String) {
        self.init(std.string(path), std.string(resolved))
    }

    /// The authored asset path as a Swift String.
    public var assetPath: String {
        return String(self.GetAssetPath())
    }

    /// The resolved asset path as a Swift String.
    public var resolvedPath: String {
        return String(self.GetResolvedPath())
    }
}

// MARK: - SdfLayer Extensions

extension SdfLayer: CustomStringConvertible {
    /// A textual representation of the layer.
    public var description: String {
        if !self.IsValid() {
            return "SdfLayer(invalid)"
        }
        return "SdfLayer(\(String(self.GetIdentifier())))"
    }
}

extension SdfLayer: CustomDebugStringConvertible {
    /// A debug textual representation of the layer.
    public var debugDescription: String {
        if !self.IsValid() {
            return "SdfLayer(invalid)"
        }
        let id = String(self.GetIdentifier())
        let display = String(self.GetDisplayName())
        let anon = self.IsAnonymous() ? ", anonymous" : ""
        let dirty = self.IsDirty() ? ", dirty" : ""
        return "SdfLayer(\(id), displayName: \(display)\(anon)\(dirty))"
    }
}

extension SdfLayer {
    /// Whether this is a valid layer reference.
    public var isValid: Bool {
        return self.IsValid()
    }

    /// The layer identifier (file path or anonymous ID).
    public var identifier: String {
        return String(self.GetIdentifier())
    }

    /// The layer's display name.
    public var displayName: String {
        return String(self.GetDisplayName())
    }

    /// Whether this is an anonymous layer.
    public var isAnonymous: Bool {
        return self.IsAnonymous()
    }

    /// Whether the layer has been modified.
    public var isDirty: Bool {
        return self.IsDirty()
    }

    /// Export the layer to a file.
    public func export(to filename: String) -> Bool {
        return self.Export(std.string(filename))
    }

    /// Create a new anonymous layer.
    public static func createAnonymous() -> SdfLayer {
        return SdfLayer.CreateAnonymous()
    }

    /// Create a new layer at the given path.
    public static func create(at path: String) -> SdfLayer {
        return SdfLayer.CreateNew(std.string(path))
    }

    /// Find or open an existing layer.
    public static func findOrOpen(_ path: String) -> SdfLayer {
        return SdfLayer.FindOrOpen(std.string(path))
    }
}

// MARK: - SdfLayerOffset Extensions

extension SdfLayerOffset: CustomStringConvertible {
    /// A textual representation of the layer offset.
    public var description: String {
        if self.IsIdentity() {
            return "SdfLayerOffset(identity)"
        }
        return "SdfLayerOffset(offset: \(self.GetOffset()), scale: \(self.GetScale()))"
    }
}

extension SdfLayerOffset: CustomDebugStringConvertible {
    /// A debug textual representation of the layer offset.
    public var debugDescription: String {
        return "SdfLayerOffset(offset: \(self.GetOffset()), scale: \(self.GetScale()), valid: \(self.IsValid()))"
    }
}

extension SdfLayerOffset: Hashable {
    /// Hashes the essential components of this value.
    public func hash(into hasher: inout Hasher) {
        hasher.combine(self.GetHash())
    }
}

extension SdfLayerOffset {
    /// Creates an identity layer offset (no transformation).
    public init() {
        self.init(0.0, 1.0)
    }

    /// Creates a layer offset with the given offset and scale.
    public init(offset: Double, scale: Double = 1.0) {
        self.init(offset, scale)
    }

    /// The time offset value.
    public var offset: Double {
        get { return self.GetOffset() }
        set { self.SetOffset(newValue) }
    }

    /// The time scale factor.
    public var scale: Double {
        get { return self.GetScale() }
        set { self.SetScale(newValue) }
    }

    /// Whether this is an identity transformation (offset=0, scale=1).
    public var isIdentity: Bool {
        return self.IsIdentity()
    }

    /// Whether this offset is valid (finite values).
    public var isValid: Bool {
        return self.IsValid()
    }

    /// The inverse of this offset.
    public var inverse: SdfLayerOffset {
        return self.GetInverse()
    }

    /// An identity layer offset (no transformation).
    public static var identity: SdfLayerOffset {
        return SdfLayerOffset()
    }

    /// Apply this offset to a time value.
    public func apply(to time: Double) -> Double {
        return self * time
    }
}

// MARK: - Common Path Constants

extension SdfPath {
    /// Common path values used throughout USD.
    public enum Common {
        /// The absolute root path "/".
        nonisolated(unsafe) public static let root = SdfPath.absoluteRoot

        /// An empty path.
        nonisolated(unsafe) public static let empty = SdfPath.empty
    }
}

File: Sources/CxxFacade/src/usd/Sdf.cpp
// Sdf.cpp - Scene Description Foundation implementation
#include "../../include/usd/Sdf.hpp"

#include <pxr/usd/sdf/path.h>
#include <pxr/usd/sdf/layer.h>
#include <pxr/usd/sdf/assetPath.h>
#include <pxr/usd/sdf/layerOffset.h>

PXR_NAMESPACE_USING_DIRECTIVE

namespace cxxfacade {
namespace sdf {

// ============================================================================
// SdfLayerOffset implementation
// ============================================================================

static inline pxr::SdfLayerOffset* asLayerOffset(void* impl) {
    return static_cast<pxr::SdfLayerOffset*>(impl);
}

SdfLayerOffset::SdfLayerOffset(double offset, double scale)
    : impl_(new pxr::SdfLayerOffset(offset, scale)) {}

SdfLayerOffset::SdfLayerOffset(const SdfLayerOffset& other)
    : impl_(new pxr::SdfLayerOffset(*asLayerOffset(other.impl_))) {}

SdfLayerOffset::SdfLayerOffset(SdfLayerOffset&& other) noexcept
    : impl_(other.impl_) {
    other.impl_ = nullptr;
}

SdfLayerOffset::~SdfLayerOffset() {
    delete asLayerOffset(impl_);
}

SdfLayerOffset& SdfLayerOffset::operator=(const SdfLayerOffset& other) {
    if (this != &other) {
        delete asLayerOffset(impl_);
        impl_ = new pxr::SdfLayerOffset(*asLayerOffset(other.impl_));
    }
    return *this;
}

SdfLayerOffset& SdfLayerOffset::operator=(SdfLayerOffset&& other) noexcept {
    if (this != &other) {
        delete asLayerOffset(impl_);
        impl_ = other.impl_;
        other.impl_ = nullptr;
    }
    return *this;
}

double SdfLayerOffset::GetOffset() const {
    if (!impl_) return 0.0;
    return asLayerOffset(impl_)->GetOffset();
}

double SdfLayerOffset::GetScale() const {
    if (!impl_) return 1.0;
    return asLayerOffset(impl_)->GetScale();
}

void SdfLayerOffset::SetOffset(double newOffset) {
    if (impl_) {
        asLayerOffset(impl_)->SetOffset(newOffset);
    }
}

void SdfLayerOffset::SetScale(double newScale) {
    if (impl_) {
        asLayerOffset(impl_)->SetScale(newScale);
    }
}

bool SdfLayerOffset::IsIdentity() const {
    if (!impl_) return true;
    return asLayerOffset(impl_)->IsIdentity();
}

bool SdfLayerOffset::IsValid() const {
    if (!impl_) return false;
    return asLayerOffset(impl_)->IsValid();
}

SdfLayerOffset SdfLayerOffset::GetInverse() const {
    if (!impl_) return SdfLayerOffset();
    return SdfLayerOffset::FromImpl(
        new pxr::SdfLayerOffset(asLayerOffset(impl_)->GetInverse()));
}

size_t SdfLayerOffset::GetHash() const {
    if (!impl_) return 0;
    return asLayerOffset(impl_)->GetHash();
}

bool SdfLayerOffset::operator==(const SdfLayerOffset& rhs) const {
    if (!impl_ && !rhs.impl_) return true;
    if (!impl_ || !rhs.impl_) return false;
    return *asLayerOffset(impl_) == *asLayerOffset(rhs.impl_);
}

bool SdfLayerOffset::operator!=(const SdfLayerOffset& rhs) const {
    return !(*this == rhs);
}

bool SdfLayerOffset::operator<(const SdfLayerOffset& rhs) const {
    if (!impl_) return rhs.impl_ != nullptr;
    if (!rhs.impl_) return false;
    return *asLayerOffset(impl_) < *asLayerOffset(rhs.impl_);
}

bool SdfLayerOffset::operator>(const SdfLayerOffset& rhs) const {
    return rhs < *this;
}

bool SdfLayerOffset::operator<=(const SdfLayerOffset& rhs) const {
    return !(rhs < *this);
}

bool SdfLayerOffset::operator>=(const SdfLayerOffset& rhs) const {
    return !(*this < rhs);
}

SdfLayerOffset SdfLayerOffset::operator*(const SdfLayerOffset& rhs) const {
    if (!impl_ || !rhs.impl_) return SdfLayerOffset();
    return SdfLayerOffset::FromImpl(
        new pxr::SdfLayerOffset(*asLayerOffset(impl_) * *asLayerOffset(rhs.impl_)));
}

double SdfLayerOffset::operator*(double rhs) const {
    if (!impl_) return rhs;
    return *asLayerOffset(impl_) * rhs;
}

SdfLayerOffset SdfLayerOffset::FromImpl(void* impl) {
    SdfLayerOffset offset;
    delete asLayerOffset(offset.impl_);
    offset.impl_ = impl;
    return offset;
}

// ============================================================================
// SdfPath implementation
// ============================================================================

static inline pxr::SdfPath* asPath(void* impl) {
    return static_cast<pxr::SdfPath*>(impl);
}

SdfPath::SdfPath() : impl_(new pxr::SdfPath()) {}

SdfPath::SdfPath(const std::string& path) : impl_(new pxr::SdfPath(path)) {}

SdfPath::SdfPath(const char* path) : impl_(new pxr::SdfPath(path)) {}

SdfPath::SdfPath(const SdfPath& other)
    : impl_(new pxr::SdfPath(*asPath(other.impl_))) {}

SdfPath::SdfPath(SdfPath&& other) noexcept : impl_(other.impl_) {
    other.impl_ = nullptr;
}

SdfPath::~SdfPath() {
    delete asPath(impl_);
}

SdfPath& SdfPath::operator=(const SdfPath& other) {
    if (this != &other) {
        delete asPath(impl_);
        impl_ = new pxr::SdfPath(*asPath(other.impl_));
    }
    return *this;
}

SdfPath& SdfPath::operator=(SdfPath&& other) noexcept {
    if (this != &other) {
        delete asPath(impl_);
        impl_ = other.impl_;
        other.impl_ = nullptr;
    }
    return *this;
}

std::string SdfPath::GetString() const {
    if (!impl_) return "";
    return asPath(impl_)->GetString();
}

const char* SdfPath::GetText() const {
    if (!impl_) return "";
    return asPath(impl_)->GetText();
}

bool SdfPath::IsEmpty() const {
    return !impl_ || asPath(impl_)->IsEmpty();
}

bool SdfPath::IsAbsolutePath() const {
    return impl_ && asPath(impl_)->IsAbsolutePath();
}

bool SdfPath::IsPrimPath() const {
    return impl_ && asPath(impl_)->IsPrimPath();
}

bool SdfPath::IsPropertyPath() const {
    return impl_ && asPath(impl_)->IsPropertyPath();
}

SdfPath SdfPath::GetParentPath() const {
    if (!impl_) return SdfPath();
    return SdfPath::FromImpl(new pxr::SdfPath(asPath(impl_)->GetParentPath()));
}

tf::TfToken SdfPath::GetName() const {
    if (!impl_) return tf::TfToken();
    return tf::TfToken::FromImpl(new pxr::TfToken(asPath(impl_)->GetName()));
}

SdfPath SdfPath::AppendChild(const tf::TfToken& childName) const {
    if (!impl_) return SdfPath();
    pxr::TfToken* pxrToken = static_cast<pxr::TfToken*>(childName.GetImpl());
    return SdfPath::FromImpl(new pxr::SdfPath(asPath(impl_)->AppendChild(*pxrToken)));
}

SdfPath SdfPath::AppendProperty(const tf::TfToken& propName) const {
    if (!impl_) return SdfPath();
    pxr::TfToken* pxrToken = static_cast<pxr::TfToken*>(propName.GetImpl());
    return SdfPath::FromImpl(new pxr::SdfPath(asPath(impl_)->AppendProperty(*pxrToken)));
}

SdfPath SdfPath::AbsoluteRootPath() {
    return SdfPath::FromImpl(new pxr::SdfPath(pxr::SdfPath::AbsoluteRootPath()));
}

SdfPath SdfPath::EmptyPath() {
    return SdfPath();
}

bool SdfPath::operator==(const SdfPath& other) const {
    if (!impl_ && !other.impl_) return true;
    if (!impl_ || !other.impl_) return false;
    return *asPath(impl_) == *asPath(other.impl_);
}

bool SdfPath::operator!=(const SdfPath& other) const {
    return !(*this == other);
}

bool SdfPath::operator<(const SdfPath& other) const {
    if (!impl_) return other.impl_ != nullptr;
    if (!other.impl_) return false;
    return *asPath(impl_) < *asPath(other.impl_);
}

SdfPath SdfPath::FromImpl(void* impl) {
    SdfPath path;
    delete asPath(path.impl_);
    path.impl_ = impl;
    return path;
}

// ============================================================================
// SdfAssetPath implementation
// ============================================================================

static inline pxr::SdfAssetPath* asAssetPath(void* impl) {
    return static_cast<pxr::SdfAssetPath*>(impl);
}

SdfAssetPath::SdfAssetPath() : impl_(new pxr::SdfAssetPath()) {}

SdfAssetPath::SdfAssetPath(const std::string& path)
    : impl_(new pxr::SdfAssetPath(path)) {}

SdfAssetPath::SdfAssetPath(const std::string& path, const std::string& resolvedPath)
    : impl_(new pxr::SdfAssetPath(path, resolvedPath)) {}

SdfAssetPath::SdfAssetPath(const SdfAssetPath& other)
    : impl_(new pxr::SdfAssetPath(*asAssetPath(other.impl_))) {}

SdfAssetPath::SdfAssetPath(SdfAssetPath&& other) noexcept : impl_(other.impl_) {
    other.impl_ = nullptr;
}

SdfAssetPath::~SdfAssetPath() {
    delete asAssetPath(impl_);
}

SdfAssetPath& SdfAssetPath::operator=(const SdfAssetPath& other) {
    if (this != &other) {
        delete asAssetPath(impl_);
        impl_ = new pxr::SdfAssetPath(*asAssetPath(other.impl_));
    }
    return *this;
}

SdfAssetPath& SdfAssetPath::operator=(SdfAssetPath&& other) noexcept {
    if (this != &other) {
        delete asAssetPath(impl_);
        impl_ = other.impl_;
        other.impl_ = nullptr;
    }
    return *this;
}

std::string SdfAssetPath::GetAssetPath() const {
    if (!impl_) return "";
    return asAssetPath(impl_)->GetAssetPath();
}

std::string SdfAssetPath::GetResolvedPath() const {
    if (!impl_) return "";
    return asAssetPath(impl_)->GetResolvedPath();
}

bool SdfAssetPath::operator==(const SdfAssetPath& other) const {
    if (!impl_ && !other.impl_) return true;
    if (!impl_ || !other.impl_) return false;
    return *asAssetPath(impl_) == *asAssetPath(other.impl_);
}

bool SdfAssetPath::operator!=(const SdfAssetPath& other) const {
    return !(*this == other);
}

SdfAssetPath SdfAssetPath::FromImpl(void* impl) {
    SdfAssetPath path;
    delete asAssetPath(path.impl_);
    path.impl_ = impl;
    return path;
}

// ============================================================================
// SdfLayer implementation
// ============================================================================

// We store a shared_ptr internally using a wrapper struct
struct SdfLayerHandle {
    pxr::SdfLayerRefPtr ptr;
    SdfLayerHandle() = default;
    explicit SdfLayerHandle(const pxr::SdfLayerRefPtr& p) : ptr(p) {}
};

static inline SdfLayerHandle* asLayerHandle(void* impl) {
    return static_cast<SdfLayerHandle*>(impl);
}

SdfLayer::SdfLayer() : impl_(new SdfLayerHandle()) {}

SdfLayer::SdfLayer(const SdfLayer& other)
    : impl_(new SdfLayerHandle(asLayerHandle(other.impl_)->ptr)) {}

SdfLayer::SdfLayer(SdfLayer&& other) noexcept : impl_(other.impl_) {
    other.impl_ = nullptr;
}

SdfLayer::~SdfLayer() {
    delete asLayerHandle(impl_);
}

SdfLayer& SdfLayer::operator=(const SdfLayer& other) {
    if (this != &other) {
        delete asLayerHandle(impl_);
        impl_ = new SdfLayerHandle(asLayerHandle(other.impl_)->ptr);
    }
    return *this;
}

SdfLayer& SdfLayer::operator=(SdfLayer&& other) noexcept {
    if (this != &other) {
        delete asLayerHandle(impl_);
        impl_ = other.impl_;
        other.impl_ = nullptr;
    }
    return *this;
}

bool SdfLayer::IsValid() const {
    return impl_ && asLayerHandle(impl_)->ptr;
}

std::string SdfLayer::GetIdentifier() const {
    if (!IsValid()) return "";
    return asLayerHandle(impl_)->ptr->GetIdentifier();
}

std::string SdfLayer::GetDisplayName() const {
    if (!IsValid()) return "";
    return asLayerHandle(impl_)->ptr->GetDisplayName();
}

bool SdfLayer::IsAnonymous() const {
    return IsValid() && asLayerHandle(impl_)->ptr->IsAnonymous();
}

bool SdfLayer::IsDirty() const {
    return IsValid() && asLayerHandle(impl_)->ptr->IsDirty();
}

bool SdfLayer::Save() const {
    if (!IsValid()) return false;
    return asLayerHandle(impl_)->ptr->Save();
}

bool SdfLayer::Export(const std::string& filename) const {
    if (!IsValid()) return false;
    return asLayerHandle(impl_)->ptr->Export(filename);
}

void SdfLayer::Clear() {
    if (IsValid()) {
        asLayerHandle(impl_)->ptr->Clear();
    }
}

SdfLayer SdfLayer::CreateAnonymous() {
    SdfLayer layer;
    asLayerHandle(layer.impl_)->ptr = pxr::SdfLayer::CreateAnonymous();
    return layer;
}

SdfLayer SdfLayer::CreateNew(const std::string& identifier) {
    SdfLayer layer;
    asLayerHandle(layer.impl_)->ptr = pxr::SdfLayer::CreateNew(identifier);
    return layer;
}

SdfLayer SdfLayer::FindOrOpen(const std::string& identifier) {
    SdfLayer layer;
    asLayerHandle(layer.impl_)->ptr = pxr::SdfLayer::FindOrOpen(identifier);
    return layer;
}

SdfLayer SdfLayer::FromImpl(void* impl) {
    SdfLayer layer;
    delete asLayerHandle(layer.impl_);
    layer.impl_ = impl;
    return layer;
}

} // namespace sdf
} // namespace cxxfacade
